dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/RateLimiter.java

```diff
@@ -0,0 +1,224 @@
+            }
+
+            long rLimit;
+            long duration;
+            Duration rLimitDuration;
+            Duration tokAddingRate;
+            String literalConstant;
+            String rateDuration;
+            String rateDurationUnitChar;
+
+            Matcher matcher = eventPattern.matcher(input);
+
+            if (!matcher.matches()) {
+                throw new UnknownFormatConversionException("input argument is not in valid format and can't be tokenize");
+            } else {
+                /**
+                 * eventPattern regex is divided into 5 groups
+                 * 0 group being the whole expression
+                 * 1 group is numeric rate limit value and will always
+                 *                          be present if exp matched
+                 * 2 is literal constant '-per-' and is optional
+                 * 3 is numeric rate duration and is optional
+                 * 4 is a char(s/m/h) and is optional
+                 */
+                rLimit = Long.parseLong(matcher.group(1)); // will always have a numeric value that fits in Long
+                literalConstant = matcher.group(2); // can be null or empty or '-per-'
+                rateDuration = matcher.group(3);// can be null or empty or numeric value
+                rateDurationUnitChar = matcher.group(4); //can be empty or char s,m,h
+            }
+
+            if (literalConstant == null || literalConstant.isEmpty()) {
+                //return rate with default duration and token adding rate
+                return new RateLimitEvent(rLimit, Constants.DEFAULT_RATE_LIMIT_DURATION, Constants.DEFAULT_TOKEN_ADDING_RATE);
+            }
+
+            // value is either going to be a valid number or null. defaulting to 1 unit(every hour/every min and so on)
+            duration = NumberUtils.isParsable(rateDuration) ? Long.parseLong(rateDuration) : 1;
+
+            switch (rateDurationUnitChar) {
+                case "s":
+                case "S":
+                    rLimitDuration = Duration.ofSeconds(duration);
+                    tokAddingRate = Duration.ofSeconds(rLimit / duration);
+                    break;
+                case "m":
+                case "M":
+                    rLimitDuration = Duration.ofMinutes(duration);
+                    tokAddingRate = Duration.ofMinutes(rLimit / duration);
+                    break;
+                case "h":
+                case "H":
+                    rLimitDuration = Duration.ofHours(duration);
+                    tokAddingRate = Duration.ofHours(rLimit / duration);
+                    break;
+                default:
+                    //should not reach here are regex will enforce char, keeping it for any unexpected use case.
+                    throw new UnknownFormatConversionException("Unexpected format encountered, can't tokenize input");
+            }
+            return new RateLimitEvent(rLimit, rLimitDuration, tokAddingRate);
```

Comment:
same note here, I think `rLimit` carries duplicate data with `rLimitDuration` and should be consolidated into one